core/src/test/scala/integration/kafka/api/DropCorruptedFilesTest.scala

```diff
@@ -183,5 +197,118 @@ class DropCorruptedFilesTest extends ZooKeeperTestHarness {
+        props.setProperty(KafkaConfig.ControlledShutdownEnableProp, "false")
+        props
+      }}
+      .map(KafkaConfig.fromProps)
+    // start servers in reverse order to ensure broker 1 becomes the controller
+    servers = serverConfigs.reverseMap(s => TestUtils.createServer(s))
+    val controllerId = TestUtils.waitUntilControllerElected(zkClient)
+    assertTrue(controllerId == 1)
+
+    // create a topic
+    val topicConfig = new Properties()
+    topicConfig.put(TopicConfig.UNCLEAN_LEADER_ELECTION_ENABLE_CONFIG, "true")
+    val expectedReplicaAssignment = Map(0 -> List(0))
+    TestUtils.createTopic(zkClient, topic, partitionReplicaAssignment = expectedReplicaAssignment, servers = servers, topicConfig = topicConfig)
+    val dataBroker = servers.find(_.config.brokerId == 0).get
+
+    val producerConfigs = new Properties()
+    producerConfigs.put(ProducerConfig.ACKS_CONFIG, "-1")
+    val producer = TestUtils.createProducer(getBrokerListStrFromServers(servers))
+    (dataBroker, producer)
+  }
+
+  @Test
+  def testCorruptedLog(): Unit = {
+    val (dataBroker, producer) = setupSingleBrokerClusterWithTopic(100)
+
+    val record = new ProducerRecord[Array[Byte], Array[Byte]](topic, partition, "key".getBytes(StandardCharsets.UTF_8),
+      "value".getBytes(StandardCharsets.UTF_8))
+    // produce 3 message
+    val totalMessages = 3
+    for (_ <- 0 until totalMessages)
+      producer.send(record).get
+
+    def segmentsOnBroker = dataBroker.replicaManager.logManager.getLog(tp).get.segments
+
+    assertEquals(totalMessages, segmentsOnBroker.numberOfSegments)
+
+    // create corruption in the middle segment
+    val middleSegment = segmentsOnBroker.get(1).get
+    val segmentFile = middleSegment.log.file()
+    val raf = new RandomAccessFile(segmentFile, "rw")
+    val mappedLogSegment = raf.getChannel.map(FileChannel.MapMode.READ_WRITE, 0, HEADER_SIZE_UP_TO_MAGIC)
+    val currentSize = mappedLogSegment.getInt(SIZE_OFFSET)
+    assertTrue(currentSize >= LegacyRecord.RECORD_OVERHEAD_V0)
+    // set the size to a value that's smaller than the smallest overhead to cause a corruption
+    mappedLogSegment.putInt(SIZE_OFFSET, 1)
+
+    dataBroker.shutdown()
+    // delete the broker's clean shutdown file, so that it tries to recover log segments upon startup
+    val logDir = dataBroker.config.logDirs(0)
+    val cleanShutdownFile = new File(logDir, Log.CleanShutdownFile)
+    cleanShutdownFile.delete()
+
+    // delete the recovery-point-offset-checkpoint file so that all log segments need to go through recovery
+    val recoveryPointCheckpointFile = new File(logDir, RecoveryPointCheckpointFile)
+    recoveryPointCheckpointFile.delete()
+
+    dataBroker.replicaManager.logManager
+    dataBroker.startup()
+
+    // verify that the segments after the corrupted are deleted and the corrupted segment is truncated to size 0
+    assertEquals(2, segmentsOnBroker.numberOfSegments)
```

Comment:
Is there a race condition here with dataBroker.startup(), or is the latter a synchronous call?

Reply:
Yes, the startup() method is a synchroneous call.